### ISSUE TITLE
Make share results distinguishable if there are more than one with the same display name

### ISSUE THREAD

**Issue by tobiasKaminsky** [nextcloud/server#24407]
- [ ] Android @tobiasKaminsky 
- [ ] Desktop @er-vin
- [ ] iOS @marinofaggiana 
- [ ] Server https://github.com/nextcloud/server/pull/23017

@juliushaertl can you summarize how and what you do?
Is this all possible via `ocs/v2.php/apps/files_sharing/api/v1/sharees` ?

**Comment by juliushaertl** [comment]
I'll write up some summary and try to add some mockups for better visualization of the various cases this should cover.

**Comment by tstreibl** [comment]
hope it's related, otherwise please ignore: 

would it be possible to enhace the webdav api's to reveal the creator of a file/directory in a shared environment? Seems you look at the api's from a similar perspective...

It's very tedious and unreliable to scan all "creation activities" with the webdav-activities api. The webdav client api tells me only my own username, since I shared the environment myself. So I can't use the OCS API to derive displayname or email of the "creator" of a file/directory. For a description of the use case behind it (automatic transfer of new content to other location and email notification about transfer), see https://help.nextcloud.com/t/get-the-creator-of-a-file-folder-in-a-shared-directory-based-on-webdav-client-api/37918. Didn't get community feedback so maybe it's not trivial (or nobody is interested)

**Comment by tobiasKaminsky** [comment]
It is a different question, but with:
```
curl --request GET \
  --url 'http://localhost/nc/ocs/v2.php/apps/files_sharing/api/v1/shares?format=json&subfiles=true&path=%2FallTypes%2Fshare%2Fincoming&include_tags=true' \
  --header 'Authorization: Basic dG9iaTpTSGZ6ZC04YkQyNC1xOUtSaS1zU2NRay03ZHh4eA==' \
  --header 'OCS-APIRequest: true' \
  --header 'format: json' \
  --header 'include_tags: true' \
  --header 'shared_with_me: false' \
```

you will get

```
{
        "id": "11",
        "share_type": 0,
        "uid_owner": "tobi",
        "displayname_owner": "Kaminsky Tobias",
        "permissions": 1,
        "can_edit": true,
        "can_delete": true,
        "stime": 1518445486,
        "parent": null,
        "expiration": null,
        "token": null,
        "uid_file_owner": "Tobi Kaminsky",
        "note": "",
        "label": null,
        "displayname_file_owner": "Tobi Kaminsky",
        "path": "\/allTypes\/share\/ShareAble",
        "item_type": "folder",
        "mimetype": "httpd\/unix-directory",
        "has_preview": false,
        "storage_id": "shared::\/allTypes\/share\/ShareAble",
        "storage": 4,
        "item_source": 246,
        "file_source": 246,
        "file_parent": 21376,
        "file_target": "\/ShareAble",
        "share_with": "admin",
        "share_with_displayname": "admin",
        "status": {
          "status": "dnd",
          "message": null,
          "icon": null,
          "clearAt": null
        },
```

Which gives you uid_file_owner and displayname_file_owner

**Comment by juliushaertl** [comment]
The general problem is that the display name of a user is sometimes not enough to uniquely identify a person especially on context of larger organizations where there might be multiple people with the same name. 

For that reason we would need some additional **unique label** being added to the displayname where this is relevant. For most cases the email addres should be a good unique identifier for this. For cases where there is no email address is set we would fall back to the user id, but that one might of course be rather cryptic e.g. in LDAP setups.

We basically have two options on when we want to show this:
- we always show it
  - makes the UI more cluttered 
  - might not be needed in most of the cases
- we only show it if there are users with the same displayname shown somewhere in the UI
  - cleaner ui
  - might still make the user not be uniquely identifiable e.g. in the list of existing shares, if the file is only shared to one user with that duplicate display name

For now https://github.com/nextcloud/server/pull/23017 basically implements the second one, if neede we can still decide later on if we always want to show it.

The following mock ups should make it a bit clearer:

### Keeping current behavior if display names are different

All matching users from the search have a unique display name so we don't show the unique label

![image](https://user-images.githubusercontent.com/3404133/95320373-ae2aef80-0899-11eb-8589-a706fa51fff9.png)

### Duplicate display names

THere are 4 users with the same display name, therefore we show the unique label. The first two have an email set, for the other two users there is no email so we fallback to the uid.

![image](https://user-images.githubusercontent.com/3404133/95320259-85a2f580-0899-11eb-8e4f-07dd521002a2.png)

### Share listing

We adapt the same behaviour to the share listing. The unique label is only shown for users where there are more entries with the same display name

![image](https://user-images.githubusercontent.com/3404133/95320643-09f57880-089a-11eb-929a-78e8703f0b93.png)


## API

With https://github.com/nextcloud/server/pull/23017 the unique label is returned as an extra field `share_with_displayname_unique` in the share listing api for each user share. This will be either the email or userid. 

For the sharee search an extra attribute is exposed for each user result as `shareWithDisplayNameUnique`.

Determining if the label should be shown or not is up to the clients.

Feedback is welcome, especially from @jancborchardt :wink:

**Comment by tstreibl** [comment]
> 
> 
> It is a different question, but with:
> 
> ```
> curl --request GET \
>   --url 'http://localhost/nc/ocs/v2.php/apps/files_sharing/api/v1/shares?format=json&subfiles=true&path=%2FallTypes%2Fshare%2Fincoming&include_tags=true' \
>   --header 'Authorization: Basic dG9iaTpTSGZ6ZC04YkQyNC1xOUtSaS1zU2NRay03ZHh4eA==' \
>   --header 'OCS-APIRequest: true' \
>   --header 'format: json' \
>   --header 'include_tags: true' \
>   --header 'shared_with_me: false' \
> ```
> 
> you will get
> 
> ```
> {
>         "id": "11",
>         "share_type": 0,
>         "uid_owner": "tobi",
>         "displayname_owner": "Kaminsky Tobias",
>         "permissions": 1,
>         "can_edit": true,
>         "can_delete": true,
>         "stime": 1518445486,
>         "parent": null,
>         "expiration": null,
>         "token": null,
>         "uid_file_owner": "Tobi Kaminsky",
>         "note": "",
>         "label": null,
>         "displayname_file_owner": "Tobi Kaminsky",
>         "path": "\/allTypes\/share\/ShareAble",
>         "item_type": "folder",
>         "mimetype": "httpd\/unix-directory",
>         "has_preview": false,
>         "storage_id": "shared::\/allTypes\/share\/ShareAble",
>         "storage": 4,
>         "item_source": 246,
>         "file_source": 246,
>         "file_parent": 21376,
>         "file_target": "\/ShareAble",
>         "share_with": "admin",
>         "share_with_displayname": "admin",
>         "status": {
>           "status": "dnd",
>           "message": null,
>           "icon": null,
>           "clearAt": null
>         },
> ```
> 
> Which gives you uid_file_owner and displayname_file_owner

Thanks a lot for this great help!!!! I was searching so long for this. Thanks and sorry for being off topic.

**Comment by jancborchardt** [comment]
Great proposal @juliushaertl, have nothing more to add. @er-vin @tobiasKaminsky @marinofaggiana is this also feasible on mobile or do you see any issues?
